### PR TITLE
Fix for blunt average calculation

### PR DESCRIPTION
**Pull request by kmahern**
## What

[[Link to story](https://dsdmoj.atlassian.net/browse/AP-3173)]

Fixed the blunt_average calculation in app/services/utilities/employment_income_monthly_equivalent_calculator.rb
Added employment payments to check_frequencies in app/services/remark_generators/orchestrator.rb in order to add unknown_frequency to remarks and trigger caseworker review

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/services/utilities/employment_income_monthly_equivalent_calculator.rb [hunk 1]
```diff
@@ -49,22 +49,26 @@ def weekly_to_monthly(value)
...
+          national_insurance_monthly_equiv: __send__(calc_method, calculation_value(calc_method, payment, "national_insurance")),
         )
       end
     end
 
     def dates
       @employment.employment_payments.map(&:date)
     end
+
+    def calculation_value(calc_method, payment, type)
```
Comment (jsugarman) [inline app/services/utilities/employment_income_monthly_equivalent_calculator.rb:70]: i'd rename the `type` arg `attribute`. clearer IMO

File: app/services/utilities/employment_income_monthly_equivalent_calculator.rb [hunk 2]
```diff
@@ -49,22 +49,26 @@ def weekly_to_monthly(value)
...
+      (@employment.employment_payments.sum(&value) / @employment.employment_payments.count).round(2)
     end
 
     def update_payments(calc_method)
       @employment.employment_payments.each do |payment|
         payment.update(
-          gross_income_monthly_equiv: __send__(calc_method, payment.gross_income),
-          tax_monthly_equiv: __send__(calc_method, payment.tax),
-          national_insurance_monthly_equiv: __send__(calc_method, payment.national_insurance),
+          gross_income_monthly_equiv: __send__(calc_method, calculation_value(calc_method, payment, "gross_income")),
```
Comment (jsugarman) [inline app/services/utilities/employment_income_monthly_equivalent_calculator.rb:59]: Should be able to use a symbol here?!

```suggestion
          gross_income_monthly_equiv: __send__(calc_method, calculation_value(calc_method, payment, :gross_income)),
```

and similarly for others below.

File: app/services/utilities/employment_income_monthly_equivalent_calculator.rb [hunk 3]
```diff
@@ -49,22 +49,26 @@ def weekly_to_monthly(value)
       (value * 52 / 12).round(2)
     end
 
-    def blunt_average(_value)
-      @employment.employment_payments.sum(&:gross_income) / 3
+    def blunt_average(value)
+      (@employment.employment_payments.sum(&value) / @employment.employment_payments.count).round(2)
```
Comment (jsugarman) [inline app/services/utilities/employment_income_monthly_equivalent_calculator.rb:53]: this is the `attribute` being passed in i believe
```suggestion
    def blunt_average(attribute)
      (@employment.employment_payments.sum(&attribute) / @employment.employment_payments.count).round(2)
```

File: app/services/utilities/employment_income_monthly_equivalent_calculator.rb [hunk 4]
```diff
@@ -49,22 +49,26 @@ def weekly_to_monthly(value)
...
         )
       end
     end
 
     def dates
       @employment.employment_payments.map(&:date)
     end
+
+    def calculation_value(calc_method, payment, type)
+      calc_method.eql?(:blunt_average) ? type.to_sym : payment.send(type)
```
[the remark this answers is not in the document]
Comment (colinbruce) [reply app/services/utilities/employment_income_monthly_equivalent_calculator.rb]: Hey Joel, I'm leaving the `.to_sym` as I believe that
```
@employment.employment_payments.sum(&value)
```
requires a symbol, and it should prevent FutureColin:tm: from accidentally sending values through at some time in the future :+1:

File: app/services/remark_generators/frequency_checker.rb [hunk 5]
```diff
@@ -2,7 +2,12 @@ module RemarkGenerators
   class FrequencyChecker < BaseChecker
     include Exemptable
 
-    def call
+    def self.call(assessment, collection, date_field = "payment_date")
+      new(assessment, collection).call(date_field)
+    end
+
+    def call(date_field = "payment_date")
+      @date_field = date_field
```
Comment (jsugarman) [inline app/services/remark_generators/frequency_checker.rb:10]: **nit** - feel free to ignore, I personally prefer `date_attribute` for naming and use of symbol `:payment_date` as default

**Review by jsugarman — APPROVED** [review]
👍   just a naming nit approving

File: app/services/remark_generators/orchestrator.rb [hunk 6]
```diff
@@ -55,6 +55,9 @@ def check_frequencies
       outgoings.group_by(&:type).each do |_type, collection|
         FrequencyChecker.call(@assessment, collection)
       end
+      assessment.employments.each do |job|
+        FrequencyChecker.call(@assessment, job.employment_payments, "date") if job.employment_payments.present?
```
Comment (jsugarman) [inline app/services/remark_generators/orchestrator.rb:59]: could use symbol
```suggestion
        FrequencyChecker.call(@assessment, job.employment_payments, :date) if job.employment_payments.present?
```

File: spec/services/remark_generators/frequency_checker_spec.rb [hunk 7]
```diff
@@ -116,5 +116,38 @@ module RemarkGenerators
...
     end
+
+    context "employment_payments" do
+      let(:gross_income_summary) { create :gross_income_summary, :with_everything }
+      let(:assessment) { gross_income_summary.assessment }
+      let(:amount) { 277.67 }
+
+      let(:collection) { employment.employment_payments }
+
+      context "regular payments" do
```
Comment (jsugarman) [inline spec/services/remark_generators/frequency_checker_spec.rb:127]:
```suggestion
      context "with regular payments" do
```

File: spec/services/remark_generators/frequency_checker_spec.rb [hunk 8]
```diff
@@ -116,5 +116,38 @@ module RemarkGenerators
...
+        let(:employment) { create :employment, :with_monthly_payments, assessment: gross_income_summary.assessment }
+
+        it "does not update the remarks class" do
+          original_remarks = assessment.remarks.as_json
+          described_class.call(assessment, collection, "date")
+          expect(assessment.reload.remarks.as_json).to eq original_remarks
+        end
+      end
+
+      context "irregular dates" do
```
Comment (jsugarman) [inline spec/services/remark_generators/frequency_checker_spec.rb:137]:
```suggestion
      context "with irregular dates" do
```

File: spec/services/remark_generators/frequency_checker_spec.rb [hunk 9]
```diff
@@ -116,5 +116,38 @@ module RemarkGenerators
         end
       end
     end
+
+    context "employment_payments" do
```
Comment (jsugarman) [inline spec/services/remark_generators/frequency_checker_spec.rb:120]: i can see these context names are consistent with the existing stuff...either way

```suggestion
    context "when checking employment_payments" do
```
or similar ^^

**Pull request merged**